### PR TITLE
Initial Classes

### PR DESCRIPTION
**Pull request by XedinUnknown**

### CODE REVIEW

**Comment by XedinUnknown** [comment]
@mecha, please use this and adapt your solution.

File: test/functional/AbstractRecursiveIteratorTest.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
+<?php
+
+namespace Dhii\Iterator\FuncTest;
+
+use Xpmock\TestCase;
+
+/**
+ * Tests {@see Dhii\Iterator\AbstractRecursiveIterator}.
+ *
+ * @TODO Search and replace tokens that begin with underscore.
```
Comment (XedinUnknown) [inline test/functional/AbstractRecursiveIteratorTest.php]: Remove pls.

File: src/AbstractRecursiveIterator.php [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    }
+
+    /**
+     * Retrieves the key for the current element of an iterable.
+     *
+     * @since [*next-version*]
+     *
+     * @param array|Traversable $iterable The iterable.
+     *
+     * @return mixed The current key.
```
Comment (XedinUnknown) [inline src/AbstractRecursiveIterator.php]: Maybe return should be `string|null`.

File: test/functional/AbstractRecursiveIteratorTest.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
+<?php
+
+namespace Dhii\Iterator\FuncTest;
+
+use Xpmock\TestCase;
+
+/**
+ * Tests {@see Dhii\Iterator\AbstractRecursiveIterator}.
```
Comment (XedinUnknown) [inline test/functional/AbstractRecursiveIteratorTest.php]: Missing backslash in absolute FQN.

File: test/unit/AbstractIteratorTest.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
+<?php
+
+namespace Dhii\Iterator\UnitTest;
+
+use Xpmock\TestCase;
+
+/**
+ * Tests {@see Dhii\Iterator\AbstractIterator}.
+ *
+ * @TODO Search and replace tokens that begin with underscore.
```
Comment (XedinUnknown) [inline test/unit/AbstractIteratorTest.php]: Remove pls.

File: test/unit/AbstractIteratorTest.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
+<?php
+
+namespace Dhii\Iterator\UnitTest;
+
+use Xpmock\TestCase;
+
+/**
+ * Tests {@see Dhii\Iterator\AbstractIterator}.
```
Comment (XedinUnknown) [inline test/unit/AbstractIteratorTest.php]: Missing backslash in absolute FQN.

File: test/functional/AbstractRecursiveIteratorTest.php [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+     * @since [*next-version*]
+     */
+    const TEST_SUBJECT_CLASSNAME = 'Dhii\\Iterator\\AbstractRecursiveIterator';
+
+    /**
+     * Creates a new instance of the test subject.
+     *
+     * @since [*next-version*]
+     *
+     * @return Dhii\Iterator\AbstractRecursiveIterator
```
Comment (XedinUnknown) [inline test/functional/AbstractRecursiveIteratorTest.php]: Missing backslash in absolute FQN.

File: test/unit/AbstractIteratorTest.php [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+     * @since [*next-version*]
+     */
+    const TEST_SUBJECT_CLASSNAME = 'Dhii\\Iterator\\AbstractIterator';
+
+    /**
+     * Creates a new instance of the test subject.
+     *
+     * @since [*next-version*]
+     *
+     * @return Dhii\Iterator\AbstractIterator
```
Comment (XedinUnknown) [inline test/unit/AbstractIteratorTest.php]: Missing backslash in absolute FQN.

File: src/AbstractRecursiveIterator.php [hunk 8]
```diff
@@ -257,6 +258,58 @@ protected function _getCurrentIterableValue(&$iterable)
     }
 
     /**
+     * Retrieves the current path.
+     *
+     * @since [*next-version*]
+     *
+     * @return array
+     */
+    protected function _getCurrentPath()
```
Comment (XedinUnknown) [inline src/AbstractRecursiveIterator.php]: Maybe `_getCurrentPathSegments()`?

File: src/AbstractRecursiveIterator.php [hunk 9]
```diff
@@ -257,6 +258,58 @@ protected function _getCurrentIterableValue(&$iterable)
...
+    }
+
+    /**
+     * Gets the path segment for a parent.
+     *
+     * @since [*next-version*]
+     *
+     * @param mixed $parent The parent iterable element.
+     *
+     * @return string
```
Comment (XedinUnknown) [inline src/AbstractRecursiveIterator.php]: Perhaps a short description?

**Review by mecha — COMMENTED** [review]
@XedinUnknown I've made a few changes to the way paths are created. A queue of path segments is now maintained and modified together with the parent stack. This had to be done since the stack might not contain enough information to deduce the path segments.

Consider this a review request 😆

**Comment by XedinUnknown** [comment]
You haz my approvals, @mecha.

**Pull request merged**